### PR TITLE
Upgrade acme.sh to use zerossl server as well as option to use Cloudflare DNS validation

### PR DESCRIPTION
**Pull request by kvaggelakos**
In this MR I've:

1. Moved acme.sh install from manual to online installation
2. Registering account with zerossl (no rate limiting)
3. Added the option to validate using DNS (cloudflare)

### CODE REVIEW

**Comment by saghul** [comment]
I don't want us to do curl | bash for acme. Can you please install it manually?

Also please make the env variables uppercase.

Leave the TODO comment too, still applies AFAICT.

**Comment by kvaggelakos** [comment]
I was trying the manual install, but somehow it seems (in my experience) that the DNS plugins were not getting installed. I'll change it back, but can you give it a try?

**Comment by saghul** [comment]
I think you need to git clone or you won't get the DNS plugins.

File: web/rootfs/etc/cont-init.d/10-config [hunk 1]
```diff
@@ -10,8 +10,12 @@ mkdir -p \
 # generate keys (maybe)
 if [[ $DISABLE_HTTPS -ne 1 ]]; then
     if [[ $ENABLE_LETSENCRYPT -eq 1 ]]; then
-        mkdir -p /config/acme.sh
-        sh /opt/acme.sh --install \
+        mkdir -p /config
+        cd /config/
+        git clone git@github.com:acmesh-official/acme.sh.git
```
Comment (saghul) [inline web/rootfs/etc/cont-init.d/10-config:15]: Rather than pulling it from git, please use https://github.com/acmesh-official/acme.sh/archive/refs/tags/3.0.6.tar.gz and `wget` which is already in the base image.

File: web/rootfs/etc/cont-init.d/10-config [hunk 2]
```diff
@@ -10,24 +10,37 @@ mkdir -p \
...
+        git checkout 3.0.6
+        cd acme.sh
+        sh ./acme.sh --install \
+            --home /config/acme.sh \
+            --accountemail $LETSENCRYPT_EMAIL \
+            --register-account \
+            --server zerossl
```
Comment (saghul) [inline web/rootfs/etc/cont-init.d/10-config:22]: Please make this configurable and don't alter acme.sh's default.

File: web/rootfs/etc/cont-init.d/10-config [hunk 3]
```diff
@@ -10,24 +10,37 @@ mkdir -p \
...
+            echo "Using standalone validation"
+            /config/acme.sh/acme.sh \
+                $STAGING \
+                --issue \
+    -           --standalone \
+                --pre-hook "if [[ -d /var/run/s6/services/nginx ]]; then s6-svc -d /var/run/s6/services/nginx; fi" \
+                --post-hook "if [[ -d /var/run/s6/services/nginx ]]; then s6-svc -u /var/run/s6/services/nginx; fi" \
+                -d $LETSENCRYPT_DOMAIN
         fi
-        export LE_WORKING_DIR="/config/acme.sh"
```
Comment (saghul) [inline web/rootfs/etc/cont-init.d/10-config:22]: Is this env var no longer necessary?

File: web/rootfs/etc/cont-init.d/10-config [hunk 4]
```diff
@@ -10,24 +10,37 @@ mkdir -p \
...
+                --post-hook "if [[ -d /var/run/s6/services/nginx ]]; then s6-svc -u /var/run/s6/services/nginx; fi" \
+                --dns dns_cf \
+                -d $LETSENCRYPT_DOMAIN
+        else
+            # TODO: move away from standalone mode to webroot mode.
+            echo "Using standalone validation"
+            /config/acme.sh/acme.sh \
+                $STAGING \
+                --issue \
+    -           --standalone \
```
Comment (saghul) [inline web/rootfs/etc/cont-init.d/10-config:39]: Stray hyphen.

File: web/rootfs/etc/cont-init.d/10-config [hunk 5]
```diff
@@ -10,24 +10,37 @@ mkdir -p \
...
+        sh ./acme.sh --install \
+            --home /config/acme.sh \
+            --accountemail $LETSENCRYPT_EMAIL \
+            --register-account \
+            --server zerossl
 
-        STAGING=""
-        if [[ $LETSENCRYPT_USE_STAGING -eq 1 ]]; then
-            STAGING="--staging"
+        if [ ! -z "$CF_Token" ]; then
```
Comment (saghul) [inline web/rootfs/etc/cont-init.d/10-config:24]: Can't you combine these 2 into a single one by having some VALIDATION variable which would be `--standalone` or `--dns dns_cf` ?

**Review by saghul — CHANGES_REQUESTED** [review]

File: web/rootfs/etc/cont-init.d/10-config [hunk 6]
```diff
@@ -10,24 +10,30 @@ mkdir -p \
...
+        wget https://github.com/acmesh-official/acme.sh/archive/refs/tags/$ACME_VERSION.tar.gz
+        tar xzvf acme.sh-$ACME_VERSION.tar.gz
+        rm acme.sh-$ACME_VERSION.tar.gz
+        mv acme.sh-$ACME_VERSION acme.sh
+        cd acme.sh
+        sh ./acme.sh --install \
+            --home /config/acme.sh \
+            --accountemail $LETSENCRYPT_EMAIL
 
-        STAGING=""
```
Comment (saghul) [inline web/rootfs/etc/cont-init.d/10-config:18]: Why did you remove this?

File: web/rootfs/etc/cont-init.d/10-config [hunk 7]
```diff
@@ -10,24 +10,30 @@ mkdir -p \
...
     if [[ $ENABLE_LETSENCRYPT -eq 1 ]]; then
+        ACME_VERSION=3.0.6
+        validation=${LETSENCRYPT_VALIDATION:-"--standalone"}
+        echo "Using validation method: $validation"
         mkdir -p /config/acme.sh
-        pushd /opt
-        sh ./acme.sh --install --home /config/acme.sh --accountemail $LETSENCRYPT_EMAIL
-        popd
+        cd /tmp/
+        wget https://github.com/acmesh-official/acme.sh/archive/refs/tags/$ACME_VERSION.tar.gz
```
Comment (saghul) [inline web/rootfs/etc/cont-init.d/10-config:18]: Can you please write this more succinctly? wget -o/tmp/... | tar so we don't need to move / rename folders.

**Review by saghul — CHANGES_REQUESTED** [review]
Almost there!

**Comment by kvaggelakos** [comment]
@saghul Thanks for the feedback. I don't really have time to test this fully. Feel free to fork or make changes as you see fit. May be easier than getting me to do it :)

**Comment by saghul** [comment]
If you can't test how do you k ow it works?

**Comment by kvaggelakos** [comment]
I did test it before all the changes. I don't actually have time to finish this. Closing for now.
